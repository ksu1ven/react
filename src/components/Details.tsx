import React from 'react';
import { useRouter } from 'next/router';
import { updateQueryParams } from '../utils/helpFunctions';
import { Animal } from '../utils/types';

function Details(props: Record<'detailsData', Animal>) {
  const { detailsData } = props;
  const router = useRouter();
  return (
    <aside
      className="relative w-1/3 bg-lime-700 text-white flex flex-col pt-40 items-center gap-6"
      data-testid="details"
    >
      <button
        className="absolute top-14 right-14 flex items-center justify-center w-16 h-16 bg-lime-300 text-4xl p-3 rounded-full"
        data-testid="cross"
        onClick={() => {
          const newParams = updateQueryParams(router.query, 'details', '');
          router.push(newParams.toString() ? '?' + newParams : '');
        }}
      >
        X
      </button>

      <div>
        <h1
          className="text-5xl font-extrabold text-lime-300 mb-10"
          data-testid="details-h1"
        >
          {detailsData.name}
        </h1>
        <ul className="text-3xl font-extrabold">
          <li>Avian: {detailsData.avian ? 'Yes' : 'No'}</li>
          <li>Canine: {detailsData.canine ? 'Yes' : 'No'}</li>
          <li>EarthAnimal: {detailsData.earthAnimal ? 'Yes' : 'No'}</li>
          <li>EarthInsect: {detailsData.earthInsect ? 'Yes' : 'No'}</li>
          <li>Feline: {detailsData.feline ? 'Yes' : 'No'}</li>
        </ul>
      </div>
    </aside>
  );
}

export default Details;
