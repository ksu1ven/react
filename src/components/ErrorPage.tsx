import { Link } from 'react-router-dom';

function ErrorPage() {
  return (
    <main className="error-occured min-h-screen grid place-content-center grow">
      <h2 className="text-5xl">Page not found :&#40;</h2>
      <Link
        className="w-fit bg-lime-700 py-3 px-10 mt-10 rounded text-white font-extrabold m-auto"
        to="/"
      >
        Back Home
      </Link>
    </main>
  );
}
export default ErrorPage;
