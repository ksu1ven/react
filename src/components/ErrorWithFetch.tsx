import React from 'react';
import { useRouter } from 'next/router';
import { updateQueryParams } from '../utils/helpFunctions';

function ErrorWithFetch(props: Record<string, string>) {
  const router = useRouter();

  return (
    <aside className="relative w-1/3 bg-lime-700 text-white p-40">
      <button
        className="absolute top-14 right-14 flex items-center justify-center w-16 h-16 bg-lime-300 text-4xl p-3 rounded-full"
        onClick={() => {
          const newParams = updateQueryParams(router.query, 'details', '');
          router.push(newParams.toString() ? '?' + newParams : '');
        }}
      >
        X
      </button>
      <h2 className="text-3xl">
        {`We can't find this ${props.param}. Problems with server or ${props.param} doesn't exist. Try again`}
      </h2>
    </aside>
  );
}
export default ErrorWithFetch;
