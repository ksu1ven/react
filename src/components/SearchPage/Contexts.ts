import { createContext } from 'react';
import { Animal } from '../../utils/types';

export const SearchValueContext = createContext<string>(
  localStorage.getItem('searchValue') || ''
);
export const SearchResultsContext = createContext<Readonly<Animal[]>>([]);
