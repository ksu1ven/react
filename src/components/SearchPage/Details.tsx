import { useSearchParams, useLoaderData } from 'react-router-dom';
import Loader from '../Loader';
import { Animal, apiResponse } from '../../utils/types';
import { useState } from 'react';
import { updateQueryParams } from '../../utils/helpFunctions';

function Details() {
  const [params, setParams] = useSearchParams();
  const [isLoading] = useState(false);
  const animal = useLoaderData() as Animal | null;
  console.log(animal);

  return params.has('details') ? (
    <aside className="relative w-1/3 bg-lime-700 text-white flex flex-col pt-40 items-center gap-6">
      {!isLoading ? (
        <>
          <button
            className="absolute top-14 right-14 flex items-center justify-center w-16 h-16 bg-lime-300 text-4xl p-3 rounded-full"
            onClick={() => {
              setParams(updateQueryParams(params, 'details', ''));
            }}
          >
            X
          </button>
          <h1 className="text-5xl font-extrabold text-lime-300">
            {animal?.name}
          </h1>
          <ul className="text-3xl font-extrabold">
            <li>Avian: {animal?.avian ? 'Yes' : 'No'}</li>
            <li>Canine: {animal?.canine ? 'Yes' : 'No'}</li>
            <li>EarthAnimal: {animal?.earthAnimal ? 'Yes' : 'No'}</li>
            <li>EarthInsect: {animal?.earthInsect ? 'Yes' : 'No'}</li>
            <li>Feline: {animal?.feline ? 'Yes' : 'No'}</li>
          </ul>
        </>
      ) : (
        <Loader />
      )}
    </aside>
  ) : (
    ''
  );
}

export async function loader(): Promise<Animal | null> {
  try {
    const animalName = new URL(location.href).searchParams.get('details');
    if (animalName) {
      const response = await fetch(
        'https://stapi.co/api/v1/rest/animal/search?pageNumber=0&pageSize=100',
        {
          method: 'POST',
          headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
          body: `${encodeURIComponent('name')}=${encodeURIComponent(
            animalName
          )}`,
        }
      );
      const json: apiResponse = await response.json();
      return json.animals[0];
    }
    return null;
  } catch {
    throw new Error('No such uid');
  }
}

export default Details;
