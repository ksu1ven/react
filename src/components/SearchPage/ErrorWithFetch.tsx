import { useSearchParams } from 'react-router-dom';

import { updateQueryParams } from '../../utils/helpFunctions';

function ErrorWithFetch(props: Record<string, string>) {
  const [params, setParams] = useSearchParams();

  return (
    <aside className="relative w-1/3 bg-lime-700 text-white p-40">
      <button
        className="absolute top-14 right-14 flex items-center justify-center w-16 h-16 bg-lime-300 text-4xl p-3 rounded-full"
        onClick={() => {
          setParams(updateQueryParams(params, 'details', ''));
        }}
      >
        X
      </button>
      <h2 className="text-3xl">
        {`We can't find this ${props.param}. Problems with server or ${props.param} doesn't exist. Try again`}
      </h2>
    </aside>
  );
}
export default ErrorWithFetch;
