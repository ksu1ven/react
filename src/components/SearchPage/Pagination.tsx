interface Props {
  pageNumber: number;
  setPageNumber: React.Dispatch<React.SetStateAction<number>>;
  totalPages: number;
}

function Pagination(props: Props) {
  const { pageNumber, setPageNumber, totalPages } = props;

  return (
    <div className="flex justify-center gap-4 mb-10">
      <button className="w-14 h-14 text-lime-700 text-4xl p-3 rounded-full text-white font-extrabold">
        {'<'}
      </button>
      <button
        className={
          pageNumber === 0
            ? 'w-14 h-14 bg-lime-300 p-3 rounded-full text-white font-extrabold'
            : 'w-14 h-14 bg-lime-700 p-3 rounded-full text-white font-extrabold'
        }
        onClick={() => {
          setPageNumber(pageNumber > 3 ? pageNumber - 2 : 1);
        }}
      >
        {pageNumber > 3 ? pageNumber - 3 : 1}
      </button>
      {totalPages > 2 && (
        <button
          className={
            pageNumber === 1
              ? 'w-14 h-14 bg-lime-300 p-3 rounded-full text-white font-extrabold'
              : 'w-14 h-14 bg-lime-700 p-3 rounded-full text-white font-extrabold'
          }
          onClick={() => {}}
        >
          {pageNumber > 3 ? pageNumber - 2 : 2}
        </button>
      )}
      {totalPages > 3 && (
        <button
          className={
            pageNumber === 2
              ? 'w-14 h-14 bg-lime-300 p-3 rounded-full text-white font-extrabold'
              : 'w-14 h-14 bg-lime-700 p-3 rounded-full text-white font-extrabold'
          }
          onClick={() => {
            setPageNumber(pageNumber > 2 ? pageNumber : 3);
          }}
        >
          {pageNumber > 3 ? pageNumber - 1 : 3}
        </button>
      )}
      {totalPages > 3 && (
        <button
          className="w-14 h-14 text-lime-700 text-4xl p-3 rounded-full text-white font-extrabold"
          onClick={() => {
            setPageNumber(pageNumber + 1);
          }}
        >
          ...
        </button>
      )}
      <button className="w-14 h-14 bg-lime-700 p-3 rounded-full text-white font-extrabold">
        {totalPages}
      </button>
      <button className="w-14 h-14 text-lime-700 text-4xl p-3 rounded-full text-white font-extrabold">
        {'>'}
      </button>
    </div>
  );
}

export default Pagination;
