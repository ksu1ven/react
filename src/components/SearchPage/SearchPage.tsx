import { useState } from 'react';
import { useSearchParams, Outlet } from 'react-router-dom';
import { useSelector } from 'react-redux';
import type { RootState } from '../../redux/store/store';
import SearchForm from './SearchForm';
import SearchResults from './SearchResults';
import Pagination from './Pagination';
import Loader from '../Loader';
import ErrorBoundary from '../ErrorBoundary';
import { Animal } from '../../utils/types';
import { updateQueryParams } from '../../utils/helpFunctions';
import SelectLimit from './Select';
import { SearchResultsContext } from './Contexts';

function SearchPage() {
  const [loading] = useState(false);
  const { totalPages } = useSelector((state: RootState) => state.pagination);

  const [params, setParams] = useSearchParams();

  const [searchResultsArray] = useState<Readonly<Animal[]>>([]);

  const [errorOccured] = useState(false);

  // useEffect(() => {
  //   setLoading(true);
  //   searchCards(searchValue, pageNumber, pageSize)
  //     .then((json: apiResponse) => {
  //       setTotalPages(json.page.totalPages);
  //       setLoading(false);
  //       setSearchResultsArray(json.animals);
  //     })
  //     .catch(() => setErrorOccured(true));
  // }, [searchValue, pageNumber, pageSize]);

  if (errorOccured) {
    throw new Error("Hello, I'm Error with server!");
  }
  return (
    <ErrorBoundary>
      <main className="relative min-h-screen flex flex-col grow">
        <section className="bg-lime-200 py-10">
          <SearchForm params={params} setParams={setParams} />
        </section>
        <section className="search-results grow">
          <SelectLimit params={params} setParams={setParams} />
          {!loading && (
            <>
              <SearchResultsContext.Provider value={searchResultsArray}>
                <SearchResults params={params} setParams={setParams} />
              </SearchResultsContext.Provider>

              {totalPages && (
                <Pagination params={params} setParams={setParams} />
              )}
            </>
          )}
        </section>
        {loading && <Loader />}
      </main>
      {params.has('details') && (
        <div
          className="fixed w-screen h-full bg-slate-500/70"
          onClick={() => {
            if (params.has('details'))
              setParams(updateQueryParams(params, 'details', ''));
          }}
        ></div>
      )}
      <Outlet />
    </ErrorBoundary>
  );
}

export default SearchPage;
