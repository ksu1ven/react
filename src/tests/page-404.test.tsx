import { describe, expect, it, vi } from 'vitest';
import '@testing-library/jest-dom';
import { render, screen } from '@testing-library/react';
import { Route, Routes, MemoryRouter } from 'react-router-dom';
import Details from '../components/SearchPage/Details';
import SearchPage from '../components/SearchPage/SearchPage';
import ErrorPage from '../components/ErrorPage';
import ErrorWithFetch from '../components/SearchPage/ErrorWithFetch';

describe('Tests for the 404 Page component', () => {
  it('404 page is displayed when navigating to an invalid route.', async () => {
    render(
      <MemoryRouter initialEntries={['/sdfghjsdfghj']}>
        <Routes>
          <Route path="/" element={<SearchPage />}>
            <Route index element={<Details />} loader={vi.fn()} />
          </Route>
          <Route path="*" element={<ErrorPage />} />
        </Routes>
      </MemoryRouter>
    );
    expect(screen.getByText('Page not found :(')).toBeInTheDocument();
    expect(screen.getByText('Back Home')).toBeInTheDocument();
  });

  it('Render errorElements (not errorPage) to get coverage 100%', async () => {
    render(
      <MemoryRouter>
        <ErrorWithFetch />
      </MemoryRouter>
    );
  });
});
