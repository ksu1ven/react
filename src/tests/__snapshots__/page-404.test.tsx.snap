// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Tests for the 404 Page component > 404 page is displayed when navigating to an invalid route. 1`] = `
<body>
  <div>
    <main
      class="error-occured min-h-screen grid place-content-center grow"
    >
      <h2
        class="text-5xl"
      >
        Page not found :(
      </h2>
      <a
        class="w-fit bg-lime-700 py-3 px-10 mt-10 rounded text-white font-extrabold m-auto"
        href="/"
      >
        Back Home
      </a>
    </main>
  </div>
</body>
`;
